vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,363 @@
+
+  m_socket.Control().IgnorableServerCertificateErrors()
+    .Append(winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult::Untrusted);
+  m_socket.Control().IgnorableServerCertificateErrors()
+    .Append(winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult::InvalidName);
+
+}
+
+WinRTWebSocketResource::WinRTWebSocketResource(const string& urlString)
+  : WinRTWebSocketResource(Uri{ Utf8ToUtf16(urlString) })
+{
+}
+
+WinRTWebSocketResource::~WinRTWebSocketResource() /*override*/
+{
+  Stop();
+}
+
+#pragma region Private members
+
+fire_and_forget WinRTWebSocketResource::PerformConnect()
+{
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    co_await m_socket.ConnectAsync(m_uri);
+
+    if (m_connectHandler)
+    {
+      m_connectHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    hr = e.code();
+  }
+
+  if (!SUCCEEDED(hr))
+  {
+    hresult_error e{ hr };
+    //OutputDebugString("WebSocket.connect failed (0x%8X) %ls\n", e);
+
+    if (m_errorHandler)
+    {
+      m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Connection });
+    }
+  }
+
+  // m_connectPerformed.Set();
+  m_connectPerformed.set_value();
+  m_connectRequested = false;
+}
+
+fire_and_forget WinRTWebSocketResource::PerformPing()
+{
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    if (m_connectRequested)
+    {
+      m_connectPerformed.get_future().wait();
+    }
+
+    m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+    string s{};
+    winrt::array_view<const uint8_t> arr(
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()),
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()) + s.length());
+    m_writer.WriteBytes(arr);
+
+    co_await m_writer.StoreAsync();
+
+    if (m_pingHandler)
+    {
+      m_pingHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    hr = e.code();
+  }
+
+  if (!SUCCEEDED(hr))
+  {
+    if (m_errorHandler)
+    {
+      m_errorHandler({ Utf16ToUtf8(hresult_error(hr).message()), ErrorType::Ping });
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformWrite()
+{
+  if (m_writeQueue.empty())
+  {
+    co_return;
+  }
+
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    if (m_connectRequested)
+    {
+      m_connectPerformed.get_future().wait();
+    }
+
+    string message;
+    bool isBinary;
+    size_t length;
+    std::tie(message, isBinary) = m_writeQueue.front();
+    m_writeQueue.pop();
+
+    if (isBinary)
+    {
+      m_socket.Control().MessageType(SocketMessageType::Binary);
+
+      auto buffer = CryptographicBuffer::DecodeFromBase64String(Utf8ToUtf16(std::move(message)));
+      length = buffer.Length();
+      m_writer.WriteBuffer(buffer);
+    }
+    else
+    {
+      m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+      length = message.size();
+      winrt::array_view<const uint8_t> arr(
```

Comment:
>uint8_t [](start = 30, length = 7)

(nit) it seems that char could be easier to understand, because this is utf8 encoded, not some arbitrary binary data

Reply:
Agreed, but I want to diverge as little as possible from what is done in the UWP WebSocket.

In the long run, we want to use a single implementation.